src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,496 @@
+                ICollection<TSource> sourceCollection = _source as ICollection<TSource>;
+                if (sourceCollection != null)
+                {
+                    sourceCollection.CopyTo(array, index);
+                    index += sourceCollection.Count;
+                }
+                else
+                {
+                    foreach (TSource item in _source)
+                    {
+                        array[index] = item;
+                        ++index;
+                    }
+                }
+
+                _appended?.CopyTo(array, index, _appCount);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
+                _prepended?.AddTo(list, _preCount);
+                list.AddRange(_source);
+                _appended?.AddToFromEnd(list, _appCount);
+                return list;
+            }
+        }
+    }
+
+    internal sealed class SharedCollection<T>
+    {
+        // Limit the number of elements stored, because the collection being shared makes it possible
+        // for a series of operations to grow the collection large, but for only one of the earlier
+        // operations to be keeping it alive.
+        private const int MaxLength = 64;
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
+
+        public bool Append(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
+                {
+                    newStore[i] = store[i];
+                }
+
+                _store = store = newStore;
+            }
+
+            store[number] = item;
+            ++_count;
+            return true;
+        }
+
+        public bool Prepend(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
+                for (int i = 0; i != store.Length; ++i)
+                {
+                    newStore[number + i] = store[i];
```

Comment:
Ditto here; maybe use `Array.Copy`? Not sure how fast it would be since the `MaxLength` of `store` appears to be 64, but for comparison `MemoryStream` uses `Buffer.BlockCopy` if the [length of the destionation > 8](https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/MemoryStream.cs#L407), and does a for-loop otherwise.